### PR TITLE
feat: add gcc skeleton

### PR DESCRIPTION
**Pull request by Nutlope**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
feature
**What is the current behavior? (You can also link to an open issue here)**
no gcc spec
**What is the new behavior (if this is a feature change)?**
gcc spec!
**Additional info:**

### CODE REVIEW

File: dev/gcc.ts [hunk 1]
```diff
@@ -0,0 +1,3229 @@
+export const completionSpec: Fig.Spec = {
+  name: "gcc",
+  description: "The default compiler for most linux distributions",
+  options: [
+    {
+      name: "--analyzer-output",
+      description:
+        "Static analyzer report output format (html|plist|plist-multi-file|plist-html|sarif|text).",
```
Comment (Monkeyank) [inline dev/gcc.ts:8]: `html|plist|plist-multi-file|plist-html|sarif|text` Make these suggestions for the arg

File: dev/gcc.ts [hunk 2]
```diff
@@ -0,0 +1,3229 @@
...
+        " Include comments from within macros in preprocessed output",
+    },
+    {
+      name: "-cl-denorms-are-zero",
+      description: "OpenCL only. Allow denormals to be flushed to zero.",
+    },
+    {
+      name: "-cl-fast-relaxed-math",
+      description:
+        " OpenCL only. Sets -cl-finite-math-only and -cl-unsafe-math-optimizations, and defines __FAST_RELAXED_MATH__.",
```
Comment (Monkeyank) [inline dev/gcc.ts:39]: Delete extra spaces at the start of strings (applies to multiple spots in the PR)

File: dev/gcc.ts [hunk 3]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-fansi-escape-codes",
+      description: " Use ANSI escape codes for diagnostics",
+    },
+    {
+      name: "-fapinotes-cache-path",
+      description:
+        "Does nothing; API notes are no longer cached separately from modules",
+      args: { name: "directory", description: "directory" },
```
Comment (Monkeyank) [inline dev/gcc.ts:221]: Add a template of `folders`

File: dev/gcc.ts [hunk 4]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-fbranch-target-identification",
+      description:
+        "Emit branch target identification instructions for indirect branch destinations",
+    },
+    {
+      name: "-fbuild-session-file",
+      description:
+        "Use the last modification time of <file> as the build session timestamp",
+      args: { name: "file", description: "file" },
```
Comment (Monkeyank) [inline dev/gcc.ts:266]: Add a `template` of `filepaths`

File: dev/gcc.ts [hunk 5]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-fcall-saved-x8",
+      description: " Make the x8 register call-saved (AArch64 only)",
+    },
+    {
+      name: "-fcall-saved-x9",
+      description: " Make the x9 register call-saved (AArch64 only)",
+    },
+    { name: "-fcaret-diagnostics", description: " " },
```
Comment (Monkeyank) [inline dev/gcc.ts:317]: Add description

File: dev/gcc.ts [hunk 6]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-fcoverage-mapping",
+      description: "Generate coverage mapping to enable code coverage analysis",
+    },
+    {
+      name: "-fcs-profile-generate",
+      description:
+        "Generate instrumented code to collect context sensitive execution counts into <directory>/default.profraw (overridden by LLVM_PROFILE_FILE env var)",
+      args: { name: "directory", description: "directory", isOptional: true },
```
Comment (Monkeyank) [inline dev/gcc.ts:364]: `template: "folders"`

File: dev/gcc.ts [hunk 7]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    { name: "-fcxx-exceptions", description: "Enable C++ exceptions" },
+    {
+      name: "-fdata-sections",
+      description: " Place each data in its own section",
+    },
+    {
+      name: "-fdebug-compilation-dir",
+      description: "The compilation directory to embed in the debug info.",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:388]: template of folders

File: dev/gcc.ts [hunk 8]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    { name: "--analyze", description: " Run the static analyzer" },
+    {
+      name: "-arcmt-migrate-emit-errors",
+      description: "Emit ARC errors even if the migrator can fix them",
+    },
+    {
+      name: "-arcmt-migrate-report-output",
+      description: "Output path for the plist report",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:19]: Add a template

File: dev/gcc.ts [hunk 9]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-ffunction-sections",
+      description: " Place each function in its own section",
+    },
+    {
+      name: "-fglobal-isel",
+      description: " Enables the global instruction selector",
+    },
+    { name: "-fgnu-inline-asm", description: "" },
```
Comment (Monkeyank) [inline dev/gcc.ts:711]: Add description

File: dev/gcc.ts [hunk 10]
```diff
@@ -0,0 +1,3229 @@
...
+      name: "-fgnu-runtime",
+      description:
+        " Generate output compatible with the standard GNU Objective-C runtime",
+    },
+    { name: "-fgnu89-inline", description: "Use the gnu89 inline semantics" },
+    {
+      name: "-fgnuc-version",
+      description:
+        "Sets various macros to claim compatibility with the given GCC version (default is 4.2.1)",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:727]: Add `default: "4.2.1"` to args

File: dev/gcc.ts [hunk 11]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-flto-jobs",
+      description:
+        "Controls the backend parallelism of -flto=thin (default of 0 means the number of threads will be derived from the number of CPUs detected)",
+      args: { name: "value", description: "value" },
+    },
+    {
+      name: "-flto",
+      description: " Set LTO mode to either 'full' or 'thin'",
+      args: { name: "value", description: "value", isOptional: true },
```
Comment (Monkeyank) [inline dev/gcc.ts:790]: Suggestions of `full` or `thin`

File: dev/gcc.ts [hunk 12]
```diff
@@ -0,0 +1,3229 @@
...
+      name: "-fmessage-length",
+      description:
+        "Format message diagnostics so that they fit within N columns",
+      args: { name: "value", description: "value" },
+    },
+    {
+      name: "-fmodule-file",
+      description:
+        "Specify the mapping of module name to precompiled module file, or load a module file if name is omitted.",
+      args: { name: "<name", description: "<name" },
```
Comment (Monkeyank) [inline dev/gcc.ts:826]: Template of `filepaths`

File: dev/gcc.ts [hunk 13]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-fmodule-map-file",
+      description: "Load this module map file",
+      args: { name: "file", description: "file" },
```
Comment (Monkeyank) [inline dev/gcc.ts:831]: Template of `filepaths`

File: dev/gcc.ts [hunk 14]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    {
+      name: "-fopenmp",
+      description: "Parse OpenMP pragmas and generate parallel code.",
+    },
+    {
+      name: "-foptimization-record-file",
+      description:
+        "Specify the output name of the file containing the optimization remarks. Implies -fsave-optimization-record. On Darwin platforms, this cannot be used with multiple -arch <arch> options.",
+      args: { name: "file", description: "file" },
```
Comment (Monkeyank) [inline dev/gcc.ts:1400]: Template of `filepaths`

File: dev/gcc.ts [hunk 15]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-fshort-wchar",
+      description: " Force wchar_t to be a short unsigned int",
+    },
+    { name: "-fshow-column", description: " " },
+    {
+      name: "-fshow-overloads",
+      description:
+        "Which overload candidates to show when overload resolution fails: best|all; defaults to all",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:1772]: Suggestions for `best` and `all`

File: dev/gcc.ts [hunk 16]
```diff
@@ -0,0 +1,3229 @@
...
+      name: "-ftrivial-auto-var-init-stop-after",
+      description:
+        "Stop initializing trivial automatic stack variables after the specified number of instances",
+      args: { name: "value", description: "value" },
+    },
+    {
+      name: "-ftrivial-auto-var-init",
+      description:
+        "Initialize trivial automatic stack variables: uninitialized (default) | pattern",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:1900]: Add 
` 
default: "uninitialized",
suggestions: ["uninitialized", "pattern"]
`

**Review by Monkeyank — CHANGES_REQUESTED** [review]
Good first pass! There's a lot of descriptions missing for options. Also add suggestions, defaults, and template suggestions (filepaths/folders) as needed!

**Review by QuiiBz — CHANGES_REQUESTED** [review]
That's such a huge spec! I reviewed the first part of it (0-1k lines), most of the issues are missing `template` so it should be very quick to resolve

File: dev/gcc.ts [hunk 17]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-mtune",
+      description:
+        "Accepted for compatibility with GCC. Currently has no effect.",
+      args: { name: "value", description: "value" },
+    },
+    {
+      name: "-MT",
+      description: " Specify name of main file output in depfile",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:2795]: Template of `filepaths`

File: dev/gcc.ts [hunk 18]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-mtls-direct-seg-refs",
+      description:
+        " Enable direct TLS access through segment registers (default)",
+    },
+    {
+      name: "-mtls-size",
+      description:
+        "Specify bit size of immediate TLS offsets (AArch64 ELF only): 12 (for 4KB) | 24 (for 16MB, default) | 32 (for 4GB) | 48 (for 256TB, needs -mcmodel=large)",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:2779]: suggestions of `12`, `24`, `32`, `48`
You can also provide descriptions for suggestions (like this one), but you don't always have to!

File: dev/gcc.ts [hunk 19]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-objcmt-returns-innerpointer-property",
+      description:
+        "Enable migration to annotate property with NS_RETURNS_INNER_POINTER",
+    },
+    {
+      name: "-objcmt-whitelist-dir-path",
+      description:
+        "Only modify files with a filename contained in the provided directory path",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:2913]: Template of `filepaths`

File: dev/gcc.ts [hunk 20]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-pipe",
+      description: " Use pipes between commands, when possible",
+    },
+    { name: "--precompile", description: "Only precompile the input" },
+    { name: "-print-effective-triple", description: "Print" },
+    {
+      name: "-print-file-name",
+      description: "Print the full library path",
+      args: { name: "file", description: "file" },
```
Comment (Monkeyank) [inline dev/gcc.ts:2940]: Template of `filepaths`

File: dev/gcc.ts [hunk 21]
```diff
@@ -0,0 +1,3229 @@
...
+    },
+    { name: "-print-targets", description: "Print the registered targets" },
+    {
+      name: "-pthread",
+      description: "Support POSIX threads in generated code",
+    },
+    {
+      name: "--ptxas-path",
+      description: "Path to ptxas (used for compiling CUDA code)",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:2977]: template

File: dev/gcc.ts [hunk 22]
```diff
@@ -0,0 +1,3229 @@
...
+    { name: "-v", description: "Show commands to run and use verbose output" },
+    {
+      name: "-Wdeprecated",
+      description:
+        "Enable warnings for deprecated constructs and define __DEPRECATED",
+    },
+    {
+      name: "-working-directory",
+      description: "Resolve file paths relative to the specified directory",
+      args: { name: "value", description: "value" },
```
Comment (Monkeyank) [inline dev/gcc.ts:3164]: Template of `filepaths` (also variadic)

**Review by Monkeyank — CHANGES_REQUESTED** [review]

**Comment by Monkeyank** [comment]
I also prob missed a lot of the `suggestions` and `templates` to add to options, so please make sure to go through all of them!

**Comment by Nutlope** [comment]
Thanks so much for the detailed review @Monkeyank and @QuiiBz! Have some higher priority stuff I'm working on rn but will address these updates later this week

File: dev/gcc.ts [hunk 23]
```diff
@@ -0,0 +1,3229 @@
...
+    {
+      name: "-ffixed-x12",
+      description: " Reserve the x12 register (AArch64/RISC-V only)",
+    },
+    {
+      name: "-ffixed-x13",
+      description: " Reserve the x13 register (AArch64/RISC-V only)",
+    },
+    {
+      name: "-ffixed-x14",
```
[the remark this answers is not in the document]
Comment (Nutlope) [reply dev/gcc.ts:562]: I guess we could but any specific advantage to doing it that way? It seems like loading the spec will take longer with running a for loop as opposed to having them hardcoded.

**Comment by Nutlope** [comment]
@QuiiBz @Monkeyank ready for your re-review. Addressed nearly all your comments besides descriptions (~50 missing descriptions, not in the --help file and not easily findable online so I think I'll just keep them blank).

File: dev/gcc.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (QuiiBz) [reply dev/gcc.ts]: We can use both, a for loop shouldn't be slow and even noticeable for the end-users.

File: dev/gcc.ts [hunk 24]
```diff
@@ -0,0 +1,3321 @@
...
+      description:
+        "Enable device-side debug info generation. Disables ptxas optimizations.",
+    },
+    {
+      name: "--cuda-path-ignore-env",
+      description: "Ignore environment variables to detect CUDA installation",
+    },
+    {
+      name: "--cuda-path",
+      description: "CUDA installation path",
```
Comment (Monkeyank) [inline dev/gcc.ts:141]: `template` of `folders` ?

File: dev/gcc.ts [hunk 25]
```diff
@@ -0,0 +1,3321 @@
...
+    },
+    {
+      name: "-dependency-dot",
+      description: "Filename to write DOT-formatted header dependencies to",
+      args: { name: "value", description: "value" },
+    },
+    {
+      name: "-dependency-file",
+      description: "Filename (or -) to write dependency output to",
+      args: { name: "value", description: "value", suggestions: ["-"] },
```
Comment (Monkeyank) [inline dev/gcc.ts:171]: Template of `filepaths`?

File: dev/gcc.ts (hunk 7 above)
Comment (Monkeyank) [reply dev/gcc.ts:418]: Template of folders?

**Review by Monkeyank — CHANGES_REQUESTED** [review]

**Comment by Nutlope** [comment]
Thanks for the additional review @Monkeyank! Addressed all your changes

**Pull request merged**